docs/design/2019-04-11-indexmerge.md

```diff
@@ -0,0 +1,284 @@
+	return nil
+}
+tableFilters := append(tableFilters, usedConditionsInOr...)
+newConsiderConditions := remove(pushDownConditions, usedConditionsInOr)
+for cond in newConsiderConditions {
+	indexPaths = buildAccessPath([]{cond}, indexInfos)
+	if indexPaths == nil {
+		tableFiltes = append(tableFilters,cond)
+		continue
+}
+indexPath := GetIndexMergePartialPath(indexPaths,indexInfos)
+partialPaths = append(partialPaths, indexPath)
+}
+if len(partialPaths) < 2 {
+	return nil
+}
+return CreateIndexMergeIntersectionPath(partialPaths, tableFilters)
+}
+
+// Now, we just use all path in partialPaths to generate a IndexMergeIntersection.
+// We also need to merge possible paths. 
+// For example: 
+//   index:  ix1(a)
+//   condition： a > 1 and a < 10
+//   we will get two partial paths and they all use index ix1. 
+// IndexMergePath.tableFilters: 
+//   <1> tableFilters
+//   <2> after merge operation, if any indexPath's tableFilter is not nil, we 
+//       should add indexPath’s tableFilter into IndexMergePath.tableFilters
+CreateIndexMergeIntersectionPath(partialPaths, tableFilters) {
+}
+
+```
+	
+### 3.3 Executor
+Graph bellow illustrates execution of IndexMerge scan. 
+
+![Execution model](./execution_model.png =400x200)
+
+Every index plan in `PhysicalIndexMergeLookUpReader` will start an `IndexWorker` to execute the IndexScan plan and send handles to AndOrWorker. AndOrWorker is responsible for doing set operations (and, or) to get final handles. Then `AndOrWoker` sends final handles to `TableWokers` to get rows from TiKV.
+
+We can take some tricks to make execution at pipeline mode without considering the order.
+
+
+(1)	IndexMergeIntersection
+
+
+ ![Execution model](./and.jpg =400x150)  
+   We take an example to explain it. Use set1 to record rowids which are returned by ix1 but not sent to tableWorker. Use set2 to record the same thing for ix2.
+If new rowid comes from ix1, first we check if it is in set2. If so, we delete it from set2 and send it to tableWorker. Otherwise, we add it into set1. For the above figure, we use the following table to show the processing.
+
+| new rowid | set1 | set2 | sent to TableWorker |
+| :------:| :------: | :------: | :------: |
+| 2(ix1) | [2] | [ ] | [ ] |
+| 1(ix2) | [2] | [1] | [ ] |
+| 5(ix1) | [2,5] |[1] | [ ] |
+| 5(ix2) | [2] | [1] | [5] |
+| 7(ix1) | [2,7] |[1] | [ ] |
+| 2(ix2) | [7] | [1] | [2] |
+
+
+
+(2)	IndexMergeUnion
+
+
+   We take a structure(we call it set) to record which rowids are accessed. If a new rowid returned by IndexScan, check if it is in set. If in it, we just skip it. Otherwise, we add it into set and send it to tableWorker. 
+
+### 3.4 Cost Model
+Cost model will consider three factors: IO, CPU, and Network. 
+
+- `IndexMergeType` = 1
+
+	IO Cost = (totalRowCount + mergedRowCount) * scanFactor
+	
+	Network Cost = (totalRowCount + mergedRowCount) * networkFactor
+	
+	Cpu Memory Cost = totalRowCount * cpuFactor + totalRowCount * memoryFactor
+	
+- `IndexMergeType` = 2

```

Comment:
`Type2` is a special case of `Type1` that `mergedRowCount` is `0`?
Maybe we can merge `Type2` into `Type1` to let this model simpler?

Reply:
yes, type1 is a special case of type2.

Reply:
sorry, type2 is a special case of type1